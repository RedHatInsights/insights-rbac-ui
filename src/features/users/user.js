import React, { Fragment, Suspense, useContext, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Outlet, useNavigationType, useParams } from 'react-router-dom';
import { useIntl } from 'react-intl';
import { Button, Label, Stack, StackItem, Text, TextContent, TextVariants } from '@patternfly/react-core';
import { TableVariant, compoundExpand } from '@patternfly/react-table';
import { Table, TableBody, TableHeader } from '@patternfly/react-table/deprecated';
import { CheckIcon, CloseIcon } from '@patternfly/react-icons';
import { useChrome } from '@redhat-cloud-services/frontend-components/useChrome';
import SkeletonTable from '@patternfly/react-component-groups/dist/dynamic/SkeletonTable';
import debounce from 'lodash/debounce';
import Section from '@redhat-cloud-services/frontend-components/Section';
import DateFormat from '@redhat-cloud-services/frontend-components/DateFormat';
import Skeleton, { SkeletonSize } from '@redhat-cloud-services/frontend-components/Skeleton';
import useAppNavigate from '../../hooks/useAppNavigate';
import { AppLink, mergeToBasename } from '../../components/navigation/AppLink';
import { RbacBreadcrumbs } from '../../components/navigation/Breadcrumbs';
import { EmptyWithAction } from '../../components/ui-states/EmptyState';
import PermissionsContext from '../../utilities/permissionsContext';
import messages from '../../Messages';
import pathnames from '../../utilities/pathnames';
import { TableToolbarView } from '../../components/tables/TableToolbarView';
import { PageLayout, PageTitle } from '../../components/layout/PageLayout';
import { fetchRoleForUser, fetchRoles } from '../../redux/roles/actions';
import { fetchUsers } from '../../redux/users/actions';
import { BAD_UUID } from '../../helpers/dataUtilities';
import { getDateFormat } from '../../helpers/stringUtilities';
import { addRolesToGroup, fetchAdminGroup } from '../../redux/groups/actions';
import { defaultSettings } from '../../helpers/pagination';
import './user.scss';

let debouncedFetch;

const User = () => {
  const intl = useIntl();
  const navigate = useAppNavigate();
  const navigationType = useNavigationType();
  const dispatch = useDispatch();
  const { username } = useParams();
  const [filter, setFilter] = useState('');
  const [expanded, setExpanded] = useState({});
  const [loadingRolesTemp, setLoadingRolesTemp] = useState(false);
  const [selectedAddRoles, setSelectedAddRoles] = useState([]);
  const chrome = useChrome();

  const selector = ({
    roleReducer: { error, roles, isLoading: isLoadingRoles, rolesWithAccess },
    userReducer: {
      users: { data },
      isUserDataLoading: isLoadingUsers,
    },
    groupReducer: { adminGroup },
  }) => ({
    adminGroup,
    roles,
    isLoadingRoles,
    rolesWithAccess,
    user: data && data.filter((user) => user.username === username)[0],
    isLoadingUsers,
    userExists: error !== BAD_UUID,
  });

  const { roles, isLoadingRoles, rolesWithAccess, user, isLoadingUsers, userExists, adminGroup } = useSelector(selector);
  const { orgAdmin, userAccessAdministrator } = useContext(PermissionsContext);
  const isAdmin = orgAdmin || userAccessAdministrator;

  const fetchRolesData = (apiProps) => dispatch(fetchRoles(apiProps));

  useEffect(() => {
    chrome.appObjectId(username);
    dispatch(fetchAdminGroup({ chrome }));
    dispatch(fetchUsers({ ...defaultSettings, limit: 0, filters: { username } }));
    fetchRolesData({ limit: 20, offset: 0, username });
    setLoadingRolesTemp(true);
    fetchRolesData({ limit: 20, offset: 0, addFields: ['groups_in'], username }).then(() => setLoadingRolesTemp(false));
    debouncedFetch = debounce(
      (limit, offset, name, addFields, username) => fetchRolesData({ limit, offset, displayName: name, addFields, username }),
      500,
    );
    return () => chrome.appObjectId(undefined);
  }, []);

  const columns = [
    {
      title: intl.formatMessage(messages.roles),
    },
    {
      title: intl.formatMessage(messages.groups),
      cellTransforms: [compoundExpand],
    },
    {
      title: intl.formatMessage(messages.permissions),
      cellTransforms: [compoundExpand],
    },
    {
      title: intl.formatMessage(messages.lastModified),
    },
  ];

  const nestedPermissionsCells = [
    intl.formatMessage(messages.application),
    intl.formatMessage(messages.resourceType),
    intl.formatMessage(messages.operation),
  ];

  const createRows = (data, username, adminGroup) =>
    data
      ? data.reduce(
          (acc, { uuid, display_name, groups_in = [], modified, accessCount }, i) => [
            ...acc,
            {
              uuid,
              cells: [
                { title: display_name, props: { component: 'th', isOpen: false } },
                { title: loadingRolesTemp ? <Skeleton size={SkeletonSize.xs} /> : groups_in.length, props: { isOpen: expanded[uuid] === 1 } },
                { title: accessCount, props: { isOpen: expanded[uuid] === 2 } },
                { title: <DateFormat type={getDateFormat(modified)} date={modified} /> },
              ],
            },
            {
              uuid: `${uuid}-groups`,
              parent: 3 * i,
              compoundParent: 1,
              cells: [
                {
                  props: { colSpan: 4, className: 'pf-m-no-padding' },
                  title:
                    groups_in?.length > 0 ? (
                      <Table
                        ouiaId="groups-in-role-nested-table"
                        aria-label="Simple Table"
                        variant={TableVariant.compact}
                        cells={[intl.formatMessage(messages.name), intl.formatMessage(messages.description), ' ']}
                        rows={groups_in.map((group) => {
                          if (!group) return null;

                          return {
                            cells: [
                              { title: <AppLink to={pathnames['group-detail'].link.replace(':groupId', group.uuid)}>{group.name}</AppLink> },
                              group.description,
                              {
                                title:
                                  !adminGroup || !group.uuid || adminGroup.uuid === group.uuid ? null : (
                                    <AppLink
                                      to={pathnames['user-add-group-roles'].link.replace(':username', username).replace(':groupId', group.uuid)}
                                      state={{ name: group.name }}
                                    >
                                      {intl.formatMessage(messages.addRoleToThisGroup)}
                                    </AppLink>
                                  ),
                                props: { className: 'pf-v5-u-text-align-right' },
                              },
                            ],
                          };
                        })}
                      >
                        <TableHeader />
                        <TableBody />
                      </Table>
                    ) : (
                      <Text className="pf-v5-u-mx-lg pf-v5-u-my-sm">
                        {loadingRolesTemp ? intl.formatMessage(messages.loading) : intl.formatMessage(messages.noGroups)}
                      </Text>
                    ),
                },
              ],
            },
            {
              uuid: `${uuid}-permissions`,
              parent: 3 * i,
              compoundParent: 2,
              cells: [
                {
                  props: { colSpan: 4, className: 'pf-m-no-padding' },
                  title:
                    rolesWithAccess && rolesWithAccess[uuid] ? (
                      rolesWithAccess[uuid].access?.length > 0 ? (
                        <Table
                          aria-label="Simple Table"
                          ouiaId="permissions-in-role-nested-table"
                          variant={TableVariant.compact}
                          cells={nestedPermissionsCells}
                          rows={rolesWithAccess[uuid].access.map((access) => ({ cells: access.permission.split(':') }))}
                        >
                          <TableHeader />
                          <TableBody />
                        </Table>
                      ) : (
                        <Text className="pf-v5-u-mx-lg pf-v5-u-my-sm">{intl.formatMessage(messages.noPermissions)}</Text>
                      )
                    ) : (
                      <SkeletonTable rows={accessCount} columns={nestedPermissionsCells} variant={TableVariant.compact} />
                    ),
                },
              ],
            },
          ],
          [],
        )
      : [];

  const onExpand = (_event, _rowIndex, colIndex, isOpen, rowData) => {
    if (!isOpen) {
      setExpanded({ ...expanded, [rowData.uuid]: colIndex });
      // Permissions
      if (colIndex === 2) {
        dispatch(fetchRoleForUser(rowData.uuid));
      }
    } else {
      setExpanded({ ...expanded, [rowData.uuid]: -1 });
    }
  };

  const breadcrumbsList = [
    { title: intl.formatMessage(messages.users), to: mergeToBasename(pathnames.users.link) },
    { title: userExists ? username : intl.formatMessage(messages.invalidUser), isActive: true },
  ];

  const toolbarButtons = () => [
    ...(isAdmin
      ? [
          <AppLink to={pathnames['add-user-to-group'].link.replace(':username', username)} key="add-user-to-group" className="rbac-m-hide-on-sm">
            <Button ouiaId="add-user-to-group-button" variant="primary" aria-label="Add user to a group">
              {intl.formatMessage(messages.addUserToGroup)}
            </Button>
          </AppLink>,
          {
            label: intl.formatMessage(messages.addUserToGroup),
            onClick: () => {
              navigate(pathnames['add-user-to-group'].link.replace(':username', username));
            },
          },
        ]
      : []),
  ];

  return (
    <Fragment>
      {userExists ? (
        <Stack>
          <StackItem>
            <PageLayout breadcrumbs={breadcrumbsList}>
              <PageTitle
                title={username}
                renderTitleTag={() =>
                  isLoadingUsers ? (
                    <Skeleton size="xs" className="rbac-c-user__label-skeleton"></Skeleton>
                  ) : (
                    <Label color={user?.is_active && 'green'}>{intl.formatMessage(user?.is_active ? messages.active : messages.inactive)}</Label>
                  )
                }
              >
                {!isLoadingUsers && user ? (
                  <Fragment>
                    <TextContent>
                      {`${intl.formatMessage(messages.orgAdministrator)}: `}
                      {user?.is_org_admin ? (
                        <CheckIcon key="yes-icon" className="pf-v5-u-mx-sm" />
                      ) : (
                        <CloseIcon key="no-icon" className="pf-v5-u-mx-sm" />
                      )}
                      {intl.formatMessage(user?.is_org_admin ? messages.yes : messages.no)}
                    </TextContent>
                    {user?.email && <Text component={TextVariants.p}>{`${intl.formatMessage(messages.email)}: ${user.email}`}</Text>}
                    {user?.username && (
                      <TextContent>
                        <Text component={TextVariants.p}>{`${intl.formatMessage(messages.username)}: ${user.username}`}</Text>
                      </TextContent>
                    )}
                  </Fragment>
                ) : null}
              </PageTitle>
            </PageLayout>
          </StackItem>
          <StackItem>
            <Section type="content" className="rbac-c-user-roles">
              <TableToolbarView
                columns={columns}
                isExpandable
                onExpand={onExpand}
                rows={createRows(roles.data, username, adminGroup)}
                data={roles.data}
                filterValue={filter}
                ouiaId="user-details-table"
                fetchData={({ limit, offset, name }) => {
                  debouncedFetch(limit, offset, name, ['groups_in'], username);
                }}
                setFilterValue={({ name }) => setFilter(name)}
                isLoading={isLoadingRoles}
                toolbarButtons={toolbarButtons}
                pagination={roles.meta}
                filterPlaceholder={intl.formatMessage(messages.roleName).toLowerCase()}
                titlePlural={intl.formatMessage(messages.roles).toLowerCase()}
                titleSingular={intl.formatMessage(messages.role).toLowerCase()}
                tableId="user"
              />
              <Suspense>
                <Outlet
                  context={{
                    // add user to group:
                    username,
                    // add group roles:
                    selectedRoles: selectedAddRoles,
                    setSelectedRoles: setSelectedAddRoles,
                    closeUrl: pathnames['user-detail'].link.replace(':username', username),
                    addRolesToGroup: (groupId, roles) => dispatch(addRolesToGroup(groupId, roles)),
                  }}
                />
              </Suspense>
            </Section>
          </StackItem>
        </Stack>
      ) : (
        <Fragment>
          <section className="pf-v5-c-page__main-breadcrumb pf-v5-u-pb-md">
            <RbacBreadcrumbs breadcrumbs={breadcrumbsList} />
          </section>
          <EmptyWithAction
            title={intl.formatMessage(messages.userNotFound)}
            description={[intl.formatMessage(messages.userNotFoundDescription, { username })]}
            actions={[
              <Button
                key="back-button"
                className="pf-v5-u-mt-xl"
                ouiaId="back-button"
                variant="primary"
                aria-label="Back to previous page"
                onClick={() => navigate(navigationType !== 'POP' ? -1 : pathnames.users.link)}
              >
                {intl.formatMessage(messages.backToPreviousPage)}
              </Button>,
            ]}
          />
        </Fragment>
      )}
    </Fragment>
  );
};

export default User;
