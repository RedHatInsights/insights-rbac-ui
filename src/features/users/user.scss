.rbac-c-user__label-skeleton {
  width: 50px;
}

.rbac-c-user-roles {
  .pf-v5-c-dropdown {
    display: block;
  }
}

// fix for cropped placeholder: https://github.com/patternfly/patternfly/issues/3449
@media only screen and (min-width: 769px) {
  input#filter-by-username.ins-c-conditional-filter[type="text"] {
    width: 230px;
  }

  .rbac-c-user-roles {
    .pf-v5-c-dropdown {
      display: none;
    }
  }
}
