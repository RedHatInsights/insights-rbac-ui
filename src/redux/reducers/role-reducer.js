import { FETCH_ROLE, FETCH_ROLES, FETCH_ROLE_FOR_USER, FETCH_ROLE_FOR_PRINCIPAL, FETCH_ROLES_FOR_WIZARD } from '../../redux/action-types';
import { defaultSettings } from '../../helpers/shared/pagination';

// Initial State
export const rolesInitialState = {
  isLoading: false,
  isRecordLoading: false,
  roles: {
    data: [],
    meta: defaultSettings,
  },
  rolesForWizard: {
    data: [],
    meta: defaultSettings,
  },
  selectedRole: {},
};

const setLoadingState = (state) => ({ ...state, isLoading: true });
const setRecordLoadingState = (state) => ({ ...state, isRecordLoading: true });
const setRole = (state, { payload }) => ({ ...state, selectedRole: payload, isRecordLoading: false });
const setRoles = (state, { payload }) => ({ ...state, roles: payload, isLoading: false });
const setRolesWithAccess = (state, { payload }) => ({
  ...state,
  rolesWithAccess: { ...state.rolesWithAccess, [payload.uuid]: payload },
  isRecordLoading: false,
});

const setRolesForWizard = (state, { payload }) => ({ ...state, rolesForWizard: payload, isWizardLoading: false });
const setWizardLoadingState = (state) => ({ ...state, isWizardLoading: true });

export default {
  [`${FETCH_ROLE}_FULFILLED`]: setRole,
  [`${FETCH_ROLE}_PENDING`]: setRecordLoadingState,
  [`${FETCH_ROLES}_FULFILLED`]: setRoles,
  [`${FETCH_ROLES}_PENDING`]: setLoadingState,
  [`${FETCH_ROLE_FOR_USER}_FULFILLED`]: setRolesWithAccess,
  [`${FETCH_ROLE_FOR_USER}_PENDING`]: setRecordLoadingState,
  [`${FETCH_ROLE_FOR_PRINCIPAL}_FULFILLED`]: setRolesWithAccess,
  [`${FETCH_ROLE_FOR_PRINCIPAL}_PENDING`]: setRecordLoadingState,
  [`${FETCH_ROLES_FOR_WIZARD}_FULFILLED`]: setRolesForWizard,
  [`${FETCH_ROLES_FOR_WIZARD}_PENDING`]: setWizardLoadingState,
};
