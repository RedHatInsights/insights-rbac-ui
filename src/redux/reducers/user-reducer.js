import { FETCH_USERS } from '../../redux/action-types';
import { defaultSettings } from '../../helpers/shared/pagination';

// Initial State
export const usersInitialState = {
  selectedUser: {},
  isUserDataLoading: false,
  users: {
    meta: defaultSettings,
  },
};

const setLoadingState = (state) => ({ ...state, isUserDataLoading: true });

const setUsers = (state, { payload }) => ({ ...state, users: payload, isUserDataLoading: false });

export default {
  [`${FETCH_USERS}_PENDING`]: setLoadingState,
  [`${FETCH_USERS}_FULFILLED`]: setUsers,
};
