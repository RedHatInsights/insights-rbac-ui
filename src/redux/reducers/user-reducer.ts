import { PaginationDefaultI, defaultSettings } from '../../helpers/shared/pagination';
import { FETCH_USERS, UPDATE_USERS_FILTERS } from '../action-types';

export interface User {
  email: string;
  external_source_id: number;
  first_name: string;
  is_active: boolean;
  is_org_admin: boolean;
  last_name: string;
  username: string;
  uuid?: number;
}

export interface UserFilters {
  username?: string;
  email?: string;
  status?: string[];
}

export interface UserStore extends Record<string, unknown> {
  selectedUser: Record<string, unknown>;
  isUserDataLoading: boolean;
  users: {
    meta: PaginationDefaultI;
    filters: UserFilters;
    pagination: PaginationDefaultI & { redirected?: boolean };
    data?: User[];
  };
}

// Initial State
export const usersInitialState: UserStore = {
  selectedUser: {},
  isUserDataLoading: false,
  users: {
    meta: defaultSettings,
    filters: {},
    pagination: { ...defaultSettings, redirected: false },
  },
};

const setLoadingState = (state: any): UserStore => ({
  ...state,
  isUserDataLoading: true,
  users: {
    ...state.users,
    pagination: {
      ...state.users.pagination,
      redirected: false,
    },
  },
});

const setUsers = (state: UserStore, { payload }: any): UserStore => ({
  ...state,
  users: { pagination: state.users?.pagination, filters: state.users?.filters, ...payload },
  isUserDataLoading: false,
});

const setFilters = (state: any, { payload }: any) => ({ ...state, users: { ...state.users, filters: payload } });

export default {
  [`${FETCH_USERS}_PENDING`]: setLoadingState,
  [`${FETCH_USERS}_FULFILLED`]: setUsers,
  [UPDATE_USERS_FILTERS]: setFilters,
};
