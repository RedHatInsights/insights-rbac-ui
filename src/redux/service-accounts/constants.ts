export const NO_DATA = 'no-data';
export const LAST_PAGE = 'last-page';
export const RESULTS = 'results';
export const ERROR = 'error';
