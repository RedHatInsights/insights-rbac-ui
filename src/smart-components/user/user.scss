.ins-c-rbac__user-label-skeleton {
  width: 50px;
}

// fix for cropped placeholder: https://github.com/patternfly/patternfly/issues/3449
 input#filter-by-username.ins-c-conditional-filter[type="text"] {
  width: 230px;
} 
