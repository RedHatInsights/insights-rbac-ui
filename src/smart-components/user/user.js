import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import debounce from 'lodash/debounce';
import { Label, Stack, StackItem } from '@patternfly/react-core';
import { TopToolbar, TopToolbarTitle } from '../../presentational-components/shared/top-toolbar';
import { TableToolbarView } from '../../presentational-components/shared/table-toolbar-view';
import { Section, DateFormat, Skeleton } from '@redhat-cloud-services/frontend-components';
import { fetchRoles, fetchRoleForUser } from '../../redux/actions/role-actions';
import { fetchUsers } from '../../redux/actions/user-actions';
import { ListLoader } from '../../presentational-components/shared/loader-placeholders';
import { defaultSettings } from '../../helpers/shared/pagination';
import './user.scss';

import { Table, TableHeader, TableBody, TableVariant, compoundExpand } from '@patternfly/react-table';

const columns = [
  'Roles',
  {
    title: 'Groups',
    cellTransforms: [compoundExpand],
  },
  {
    title: 'Permissions',
    cellTransforms: [compoundExpand],
  },
  {
    title: 'Last commit',
  },
];

let debouncedFetch;

const User = ({
  match: {
    params: { username },
  },
  fetchRoles,
  fetchRoleForUser,
  fetchUsers,
  roles,
  isLoading,
  rolesWithAccess,
  user,
}) => {
  useEffect(() => {
    fetchUsers({ ...defaultSettings, limit: 0, username });
  }, []);

  const [filter, setFilter] = useState('');
  const [expanded, setExpanded] = useState({});

  const createRows = (data) => {
    return data
      ? data.reduce(
          (acc, { uuid, name, groups_in = [], modified, accessCount }, i) => [
            ...acc,
            {
              uuid,
              cells: [
                { title: name, props: { component: 'th', isOpen: false } },
                { title: `${groups_in.length}`, props: { isOpen: expanded[uuid] === 1 } },
                { title: accessCount, props: { isOpen: expanded[uuid] === 2 } },
                { title: <DateFormat type="exact" date={modified} /> },
              ],
            },
            {
              uuid: `${uuid}-groups`,
              parent: 3 * i,
              compoundParent: 1,
              cells: [
                {
                  props: { colSpan: 4, className: 'pf-m-no-padding' },
                  title: (
                    <Table
                      aria-label="Simple Table"
                      variant={TableVariant.compact}
                      cells={['Name', 'Description']}
                      rows={groups_in.map((g) => ({ cells: [{ title: <Link to={`/groups/detail/${g.uuid}`}>{g.name}</Link> }, g.description] }))}
                    >
                      <TableHeader />
                      <TableBody />
                    </Table>
                  ),
                },
              ],
            },
            {
              uuid: `${uuid}-groups`,
              parent: 3 * i,
              compoundParent: 2,
              cells: [
                {
                  props: { colSpan: 4, className: 'pf-m-no-padding' },
                  title:
                    rolesWithAccess && rolesWithAccess[uuid] ? (
                      <Table
                        aria-label="Simple Table"
                        variant={TableVariant.compact}
                        cells={['Application', 'Resource type', 'Operation']}
                        rows={rolesWithAccess[uuid].access.map((access) => ({ cells: access.permission.split(':') }))}
                      >
                        <TableHeader />
                        <TableBody />
                      </Table>
                    ) : (
                      <ListLoader />
                    ),
                },
              ],
            },
          ],
          []
        )
      : [];
  };

  useEffect(() => {
    fetchRoles({ limit: 20, offset: 0, addFields: ['groups_in'], username });
    debouncedFetch = debounce((limit, offset, name, addFields, username) => fetchRoles({ limit, offset, name, addFields, username }), 500);
  }, []);

  const onExpand = (_event, _rowIndex, colIndex, isOpen, rowData) => {
    if (!isOpen) {
      setExpanded({ ...expanded, [rowData.uuid]: colIndex });
      // Permissions
      if (colIndex === 2) {
        fetchRoleForUser(rowData.uuid);
      }
    } else {
      setExpanded({ ...expanded, [rowData.uuid]: -1 });
    }
  };

  const breadcrumbsList = () => [
    { title: 'Users', to: '/users' },
    { title: username, isActive: true },
  ];

  return (
    <Stack>
      <StackItem>
        <TopToolbar paddingBottm={false} breadcrumbs={breadcrumbsList()}>
          <TopToolbarTitle
            title={username}
            renderTitleTag={() =>
              user && !isLoading ? (
                <Label color={user?.is_active && 'green'}>{user?.is_active ? 'Active' : 'Inactive'}</Label>
              ) : (
                <Skeleton size="xs" className="ins-c-rbac__user-label-skeleton"></Skeleton>
              )
            }
            description={`${username}'s roles, groups and permissions.`}
          />
        </TopToolbar>
      </StackItem>
      <StackItem>
        <Section type="content" id={'user-detail'}>
          <TableToolbarView
            columns={columns}
            isCompact={false}
            isExpandable={true}
            onExpand={onExpand}
            createRows={createRows}
            data={roles.data}
            filterValue={filter}
            fetchData={({ limit, offset, name }) => {
              debouncedFetch(limit, offset, name, ['groups_in'], username);
            }}
            setFilterValue={({ name }) => setFilter(name)}
            isLoading={isLoading}
            pagination={roles.meta}
            filterPlaceholder="role name"
            titlePlural="roles"
            titleSingular="role"
          />
        </Section>
      </StackItem>
    </Stack>
  );
};

User.propTypes = {
  match: PropTypes.object,
  fetchRoles: PropTypes.func,
  fetchRoleForUser: PropTypes.func,
  fetchUsers: PropTypes.func,
  roles: PropTypes.object,
  isLoading: PropTypes.bool,
  rolesWithAccess: PropTypes.object,
  user: PropTypes.object,
};

const mapStateToProps = (
  {
    roleReducer: { roles, isLoading, rolesWithAccess },
    userReducer: {
      users: { data },
    },
  },
  {
    match: {
      params: { username },
    },
  }
) => ({
  roles,
  isLoading,
  rolesWithAccess,
  user: data && data.filter((user) => user.username === username)[0],
});
const mapDispatchToProps = (dispatch) => ({
  fetchRoles: (apiProps) => dispatch(fetchRoles(apiProps)),
  fetchRoleForUser: (uuid) => dispatch(fetchRoleForUser(uuid)),
  fetchUsers: (apiProps) => dispatch(fetchUsers(apiProps)),
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(User));
