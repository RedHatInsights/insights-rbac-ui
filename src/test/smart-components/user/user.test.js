import React from 'react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import promiseMiddleware from 'redux-promise-middleware';
import { rolesInitialState } from '../../../redux/reducers/role-reducer';
import { usersInitialState } from '../../../redux/reducers/user-reducer';
import user from '../../../smart-components/user/user';
import { mock } from '../../__mocks__/apiMock';
import { RBAC_API_BASE } from '../../../utilities/constants';

describe('<User />', () => {
  const middlewares = [promiseMiddleware];
  let mockStore;
  let enhanceState;
  let initialState;

  beforeEach(() => {
    enhanceState = {
      data: [{ username: 'epacific-insights', email: 'user@redhat.com', first_name: 'User', last_name: 'User', is_active: true }],
      meta: {
        count: 39,
        limit: 10,
        offset: undefined,
      },
    };
    mockStore = configureStore(middlewares);
    initialState = {
      userReducer: { ...usersInitialState, users: enhanceState },
      roleReducer: {
        ...rolesInitialState,
      },
    };
  });

  afterEach(() => {
    mock.reset();
  });

  it('should render user', async () => {
    mock.onGet(`${RBAC_API_BASE}/principals/?limit=0&offset=0&usernames=epacific-insights&sort_order=asc`).reply(200, {});
    mock.onGet(`${RBAC_API_BASE}/roles/?limit=20&offset=0&add_fields=groups_in&username=epacific-insights`).reply(200, {});
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <Provider store={mockStore(initialState)}>
          <MemoryRouter initialEntries={['/users/detail/epacific-insights']}>
            <Route path="/users/detail/:username" component={user} />
          </MemoryRouter>
        </Provider>
      );
    });
    wrapper.update();
    expect(wrapper.find(user)).toHaveLength(1);
  });
});
