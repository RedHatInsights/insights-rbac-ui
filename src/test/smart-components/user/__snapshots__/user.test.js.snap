// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User /> should render user 1`] = `
<div>
  <div
    class="pf-v5-l-stack"
  >
    <div
      class="pf-v5-l-stack__item"
    >
      <section
        class="pf-v5-c-page__main-breadcrumb"
      >
        <nav
          aria-label="Breadcrumb"
          class="pf-v5-c-breadcrumb"
          data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
          data-ouia-component-type="PF5/Breadcrumb"
          data-ouia-safe="true"
        >
          <ol
            class="pf-v5-c-breadcrumb__list"
            role="list"
          >
            <li
              class="pf-v5-c-breadcrumb__item"
            >
              <a
                class=""
                href="/iam/user-access/users"
              >
                Users
              </a>
            </li>
            <li
              class="pf-v5-c-breadcrumb__item"
            >
              <span
                class="pf-v5-c-breadcrumb__item-divider"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
              epacific-insights
            </li>
          </ol>
        </nav>
      </section>
      <section
        class="rbac-page-header pf-v5-l-page-header pf-v5-c-page-header pf-v5-l-page__main-section pf-v5-c-page__main-section pf-m-light"
        data-ouia-component-id="OUIA-Generated-RHI/Header-true-1"
        data-ouia-component-type="RHI/Header"
        data-ouia-safe="true"
        widget-type="InsightsPageHeader"
      >
        <div
          class="pf-v5-l-flex"
        >
          <div
            class="pf-v5-u-mb-sm"
          >
            <div
              class="pf-v5-l-flex pf-m-justify-content-space-between"
            >
              <div
                class="pf-v5-u-flex-grow-1"
              >
                <h1
                  class="pf-v5-c-title pf-m-2xl rbac-page-header__title"
                  data-ouia-component-id="OUIA-Generated-RHI/Header-true-2"
                  data-ouia-component-type="RHI/Header"
                  data-ouia-safe="true"
                  widget-type="InsightsPageHeaderTitle"
                >
                  epacific-insights
                </h1>
              </div>
            </div>
          </div>
          <div
            class="pf-m-align-self-center"
          >
            <span
              class="pf-v5-c-label pf-m-green"
            >
              <span
                class="pf-v5-c-label__content"
              >
                <span
                  class="pf-v5-c-label__text"
                >
                  Active
                </span>
              </span>
            </span>
          </div>
        </div>
        <div
          class="pf-v5-c-content"
        >
          Org. Administrator: 
          <svg
            aria-hidden="true"
            class="pf-v5-svg pf-v5-u-mx-sm"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 730 1024"
            width="1em"
          >
            <path
              d="M725.499315,767.757345 L469.242169,511.500499 L725.499315,255.243653 C729.196254,251.545719 731.096195,247.248119 731,242.4498 C731,237.653479 728.997365,233.35588 725.299427,229.558002 L647.542773,151.701495 C643.743891,147.903617 639.446287,146 634.64996,146 C629.852635,146 625.55503,147.803673 621.857092,151.500607 L365.5,407.857398 L109.241857,151.500607 C105.544917,147.803673 101.246314,145.904733 96.4499876,146 C91.6526619,146.104622 87.3550577,148.002562 83.5571748,151.701495 L5.7005771,229.457058 C1.90269428,233.255936 0,237.553535 0,242.350855 C0,247.148175 1.80175055,251.444775 5.50068853,255.143709 L261.857779,511.500499 L5.50068853,767.757345 C1.80175055,771.454279 -0.0961914123,775.752878 0,779.750645 C0.103697157,784.547965 2.00263857,788.845564 5.7005771,792.642443 L83.4572306,871.299502 C87.2541139,875.096381 91.5527176,877 96.3500433,877 C101.147369,877 105.444973,875.197325 109.142912,871.499391 L365.400058,615.241545 L621.657203,871.499391 C625.355142,875.197325 629.652746,877.095265 634.449072,877 C639.246398,877 643.545002,874.997437 647.342885,871.299502 L725.099538,793.54294 C728.896421,789.745062 730.796362,785.446463 730.796362,780.649143 C731.096195,775.752878 729.196254,771.454279 725.499315,767.757345"
            />
          </svg>
          No
        </div>
        <p
          class=""
          data-ouia-component-id="OUIA-Generated-Text-1"
          data-ouia-component-type="PF5/Text"
          data-ouia-safe="true"
          data-pf-content="true"
        >
          Email: user@redhat.com
        </p>
        <div
          class="pf-v5-c-content"
        >
          <p
            class=""
            data-ouia-component-id="OUIA-Generated-Text-2"
            data-ouia-component-type="PF5/Text"
            data-ouia-safe="true"
            data-pf-content="true"
          >
            Username: epacific-insights
          </p>
        </div>
      </section>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <section
        class="rbac-c-user-roles ins-l-content"
      >
         
        <div>
          <div
            class="pf-v5-c-empty-state pf-m-sm"
          >
            <div
              class="pf-v5-c-empty-state__content"
            >
              <div
                class="pf-v5-c-empty-state__header"
              >
                <div
                  class="pf-v5-c-empty-state__icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                    />
                  </svg>
                </div>
                <div
                  class="pf-v5-c-empty-state__title"
                >
                  <h4
                    class="pf-v5-c-empty-state__title-text"
                  >
                    Configure {items}
                  </h4>
                </div>
              </div>
              <div
                class="pf-v5-c-empty-state__body pf-v5-u-mb-md"
              >
                To configure user access to applications
                 
                <br />
                create at least one {item}
                 
                <br />
              </div>
              <div
                class="pf-v5-c-empty-state__footer"
              />
            </div>
          </div>
        </div>
         
      </section>
    </div>
  </div>
</div>
`;
